node/core/backing/src/lib.rs

```diff
@@ -0,0 +1,1720 @@
+		}
+
+		res
+	}
+
+	/// Check if there have happened any new misbehaviors and issue necessary messages.
+	async fn issue_new_misbehaviors(&mut self) -> Result<(), Error> {
+		let mut reports = Vec::new();
+
+		for (k, v) in self.table.get_misbehavior().iter() {
+			if !self.reported_misbehavior_for.contains(k) {
+				self.reported_misbehavior_for.insert(*k);
+
+				let f = FromTableMisbehavior {
+					id: *k,
+					report: v.clone(),
+					signing_context: self.table_context.signing_context.clone(),
+					key: self.table_context.validators[*k as usize].clone(),
+				};
+
+				if let Ok(report) = MisbehaviorReport::try_from(f) {
+					let message = ProvisionerMessage::ProvisionableData(
+						ProvisionableData::MisbehaviorReport(self.parent, report)
+					);
+
+					reports.push(message);
+				}
+			}
+		}
+
+		for report in reports.drain(..) {
+			self.send_to_provisioner(report).await?
+		}
+
+		Ok(())
+	}
+
+	/// Import a statement into the statement table and return the summary of the import.
+	async fn import_statement(
+		&mut self,
+		statement: &SignedFullStatement,
+	) -> Result<Option<TableSummary>, Error> {
+		let stmt = primitive_statement_to_table(statement);
+
+		let summary = self.table.import_statement(&self.table_context, stmt);
+
+		self.issue_new_misbehaviors().await?;
+
+		return Ok(summary);
+	}
+
+	async fn process_msg(&mut self, msg: CandidateBackingMessage) -> Result<(), Error> {
+		match msg {
+			CandidateBackingMessage::Second(_, candidate, pov) => {
+				// If the message is a `CandidateBackingMessage::Second`, sign and dispatch a
+				// Seconded statement only if we have not seconded any other candidate and
+				// have not signed a Valid statement for the requested candidate.
+				match self.seconded {
+					// This job has not seconded a candidate yet.
+					None => {
+						let candidate_hash = candidate.hash();
+
+						if !self.issued_validity.contains(&candidate_hash) {
+							if let Ok(ValidationResult::Valid) = self.validate_and_second(
+								candidate,
+								pov,
+							).await {
+								self.seconded = Some(candidate_hash);
+							}
+						}
+					}
+					// This job has already seconded a candidate.
+					Some(_) => {}
+				}
+			}
+			CandidateBackingMessage::Statement(_, statement) => {
+				self.check_statement_signature(&statement)?;
+				self.maybe_validate_and_import(statement).await?;
+			}
+			CandidateBackingMessage::GetBackedCandidates(_, tx) => {
+				let backed = self.get_backed();
+
+				tx.send(backed).map_err(|_| oneshot::Canceled)?;
+			}
+		}
+
+		Ok(())
+	}
+
+	/// Kick off validation work and distribute the result as a signed statement.
+	async fn kick_off_validation_work(
+		&mut self,
+		summary: TableSummary,
+	) -> Result<ValidationResult, Error> {
+		let candidate = self.table.get_candidate(&summary.candidate).ok_or(Error::CandidateNotFound)?;
+		let candidate = candidate.clone();
+		let descriptor = candidate.to_descriptor();
```

Comment:
`clone` should not be necessary with `to_descriptor`

Reply:
Conversion `to_descriptor` is not the reason.
`clone` is happening because `table.get_candidate` returns an `Option<&Candidate>` and we have to clone it to pass it by value into the validation.